File: backend/app/schemas/possession.py
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Optional, List
from pydantic import BaseModel, ConfigDict, Field


class ComplianceCheckItem(BaseModel):
    check_name: str
    is_satisfied: bool
    status_label: str
    details: str


class PossessionCreate(BaseModel):
    project_id: uuid.UUID
    parcel_id: uuid.UUID
    award_id: Optional[uuid.UUID] = None
    possession_date: date
    possession_type: str = Field(default="SECTION_38_REGULAR", description="SECTION_38_REGULAR or SECTION_40_URGENCY_CLAUSE")
    is_encumbrance_free: bool = True
    urgency_justification: Optional[str] = Field(default=None, description="Mandatory if Section 40 urgency clause is invoked")
    remarks: Optional[str] = None


class PossessionStatusUpdate(BaseModel):
    status: str = Field(..., description="Target status: SCHEDULED, TAKEN, DISPUTED, CANCELLED")
    remarks: Optional[str] = None


class PossessionListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    possession_reference: str
    project_id: uuid.UUID
    project_title: str
    project_code: str
    parcel_id: uuid.UUID
    khasra_number: str
    village_name: str
    district_name: str
    possession_date: date
    possession_type: str
    status: str
    is_encumbrance_free: bool
    taken_by_officer_name: str
    handed_over_by_officer_name: str
    created_at: datetime


class PossessionDetail(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    possession_reference: str
    status: str
    project_id: uuid.UUID
    project_title: str
    project_code: str
    district_name: str
    state_name: str
    parcel_id: uuid.UUID
    khasra_number: str
    khata_number: str
    village_name: str
    acquired_area_sqm: Decimal
    area_acres: Decimal
    land_type: str
    award_id: Optional[uuid.UUID] = None
    award_number: Optional[str] = None
    possession_date: date
    possession_type: str
    is_encumbrance_free: bool
    possession_certificate_doc_id: Optional[uuid.UUID] = None
    certificate_number: Optional[str] = None
    taken_by_agency_officer_id: uuid.UUID
    taken_by_officer_name: str
    taken_by_organization: Optional[str] = None
    handed_over_by_cala_id: uuid.UUID
    handed_over_by_officer_name: str
    handed_over_by_designation: Optional[str] = None
    remarks: Optional[str] = None
    prerequisite_checks: List[ComplianceCheckItem] = []
    created_at: datetime
    updated_at: datetime

File: frontend/src/lib/types/possession.ts
export interface ComplianceCheckItem {
  check_name: string;
  is_satisfied: boolean;
  status_label: string;
  details: string;
}

export interface PossessionListItem {
  id: string;
  possession_reference: string;
  project_id: string;
  project_title: string;
  project_code: string;
  parcel_id: string;
  khasra_number: string;
  village_name: string;
  district_name: string;
  possession_date: string;
  possession_type: string;
  status: string;
  is_encumbrance_free: boolean;
  taken_by_officer_name: string;
  handed_over_by_officer_name: string;
  created_at: string;
}

export interface PossessionDetail {
  id: string;
  possession_reference: string;
  status: string;
  project_id: string;
  project_title: string;
  project_code: string;
  district_name: string;
  state_name: string;
  parcel_id: string;
  khasra_number: string;
  khata_number: string;
  village_name: string;
  acquired_area_sqm: number | string;
  area_acres: number | string;
  land_type: string;
  award_id?: string | null;
  award_number?: string | null;
  possession_date: string;
  possession_type: string;
  is_encumbrance_free: boolean;
  possession_certificate_doc_id?: string | null;
  certificate_number?: string | null;
  taken_by_agency_officer_id: string;
  taken_by_officer_name: string;
  taken_by_organization?: string | null;
  handed_over_by_cala_id: string;
  handed_over_by_officer_name: string;
  handed_over_by_designation?: string | null;
  remarks?: string | null;
  prerequisite_checks: ComplianceCheckItem[];
  created_at: string;
  updated_at: string;
}

export interface PossessionFilterParams {
  project_id?: string;
  status?: string;
  search?: string;
  page?: number;
  page_size?: number;
}
